### DISCUSSION TITLE
[HOOK | IMPROVEMENT] add hook actionCartGetOrderTotalAfter on Cart->getOrderTotal

### DISCUSSION THREAD

**Discussion by CySSoO** [PrestaShop/PrestaShop#41077] [Ideas]
#### Is your feature request related to a problem?

For some reasons, payment modules need to change the cart total, especially for **partial payments** that would not use discounts.
Being able to increase or decrease the cart total would allow modules to make reductions or increases on the fly, and unblock partial payment on payment modules..

#### Describe the solution you'd like

Actually i'm using an override to increase or decrease cart total. Implementing a hook on an override was the only way to change cart total depending on specific payment module.
The hook must have in return parameters _reduction_ or _increase_, corresponding to the increase or reduction of the total of the cart.
The order can be rectified by the module once finalized, in particular on the displayOrderConfirmation hook.

```
// Start override
$hook = Hook::exec(
    'actionCartGetOrderTotalAfter',
    array(
        'order_total' => $value,
        'cart' => $this,
        'context' => Context::getContext(),
        'withTaxes' => $withTaxes,
        'type' => $type,
        'products' => $products,
        'id_carrier' => $id_carrier,
        'use_cache' => $use_cache,
        'keepOrderPrices' => $keepOrderPrices
    ),
    null,
    true
);
if (is_array($hook) && count($hook) > 0) {
    foreach ($hook as $module_cart_value) {
        if (isset($module_cart_value['reduction'])
            && (float)$module_cart_value['reduction'] > 0
        ) {
            $value -= $module_cart_value['reduction'];
        }
        if (isset($module_cart_value['increase'])
            && (float)$module_cart_value['increase'] > 0
        ) {
            $value += $module_cart_value['increase'];
        }
    }
}
// End override
// ROUND AND RETURN
return Tools::ps_round($value, $computePrecision);
```
#### Alternatives you've considered

Overriding method using a module. Other hooks and overrides were tested without success.

#### Additional context

Orders **must** be rewritten after being saved.

**Comment by florine2623** [comment]
Hello @CySSoO ,

Thank you for your suggestion. The Developer Team will take it into consideration for future developments. 

Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know!

Thank you

**Comment by Progi1984** [comment]
@CySSoO After thinking, your hook could do the job directly.


```
// Start override
$value = Hook::exec(
    'actionCartGetOrderTotalAfter',
    array(
        'order_total' => $value,
        'cart' => $this,
        'context' => Context::getContext(),
        'withTaxes' => $withTaxes,
        'type' => $type,
        'products' => $products,
        'id_carrier' => $id_carrier,
        'use_cache' => $use_cache,
        'keepOrderPrices' => $keepOrderPrices,
        'cartTotal' => $value,
    ),
);

return Tools::ps_round($value, $computePrecision);
```
